### PR TITLE
Integration test with UTF-8 char

### PR DESCRIPTION
**Pull request by Nexucis**
Hello,

I'm coming with the tests as we discussed in the PR #482. The problem is only when we create an index and an alias with some UTF-8 characters and after that we try to get the index by the alias or we want to verify if the given alias exists.

As a workaround, we can urlencode the alias before passed it to the client : 

```php
$params = array(
            'name' => urlencode($alias)
);
// return true even with UTF-8 char
$this->assertEquals($alias . '_v1', array_keys(self::$client->indices()->getAlias($params))[0]);
```

I run the following test with ElasticSearch 2.4 and 5.5.2

### CODE REVIEW

**Comment by Nexucis** [comment]
In fact, with the new test `testPutAlias`, we can see that the problem is with the parameter `name`. The parameter `index` seems to work properly :)

**Comment by Nexucis** [comment]
ping @polyfractal :)

File: tests/Elasticsearch/Tests/ClientIntegrationTests.php [hunk 1]
```diff
@@ -1,10 +1,11 @@
 <?php
 
-declare(strict_types = 1);
+declare(strict_types=1);
```
Comment (mhujer) [inline tests/Elasticsearch/Tests/ClientIntegrationTests.php]: should not be changed in the PR

File: tests/Elasticsearch/Tests/ClientIntegrationTests.php [hunk 2]
```diff
@@ -18,9 +19,32 @@
...
+        $param = [
+            'index' => '_all'
+        ];
+        self::$client->indices()->delete($param);
+    }
+
     public function testCustomQueryParams()
     {
-        $client = Elasticsearch\ClientBuilder::create()->setHosts([$_SERVER['ES_TEST_HOST']])->build();
+        self::$client = Elasticsearch\ClientBuilder::create()->setHosts([$_SERVER['ES_TEST_HOST']])->build();
```
Comment (mhujer) [inline tests/Elasticsearch/Tests/ClientIntegrationTests.php]: Is this change necessary? I think that it is better if the tests do not depend on static attribute.

File: tests/Elasticsearch/Tests/ClientIntegrationTests.php [hunk 3]
```diff
@@ -30,8 +54,85 @@ public function testCustomQueryParams()
...
+     * @dataProvider aliasDataProvider
+     */
+    public function testFindIndexByAlias($alias)
+    {
+        $this->createIndexWithAlias($alias);
+
+        $params = array(
+            'name' => $alias
+        );
+        $this->assertEquals($alias . '_v1', array_keys(self::$client->indices()->getAlias($params))[0]);
```
Comment (mhujer) [inline tests/Elasticsearch/Tests/ClientIntegrationTests.php]: `$this->assertSame()` should be used insted of `assertEquals` (there are more occurences in the PR)

`assertSame` is like `===`
`assertEquals` is like `==`

File: tests/Elasticsearch/Tests/ClientIntegrationTests.php [hunk 4]
```diff
@@ -30,8 +54,85 @@ public function testCustomQueryParams()
...
+    }
+
+    /**
+     * @dataProvider aliasDataProvider
+     */
+    public function testPutAlias($alias)
+    {
+
```
Comment (mhujer) [inline tests/Elasticsearch/Tests/ClientIntegrationTests.php]: unnecessary newline

File: tests/Elasticsearch/Tests/ClientIntegrationTests.php [hunk 5]
```diff
@@ -30,8 +54,85 @@ public function testCustomQueryParams()
...
+    public function aliasDataProvider()
+    {
+        return [
+            'latin-char' => ['myindextest'],
+            'utf-8-char' => ['⿇⽸⾽']
+        ];
+    }
+
+    private function createIndexWithAlias($alias)
```
Comment (mhujer) [inline tests/Elasticsearch/Tests/ClientIntegrationTests.php]: there can be a `string` typehint

File: tests/Elasticsearch/Tests/ClientIntegrationTests.php [hunk 6]
```diff
@@ -30,8 +54,85 @@ public function testCustomQueryParams()
...
+            'body' => array(
+                'aliases' => array(
+                    $alias => new stdClass()
+                )),
+        );
+
+        self::$client->indices()->create($params);
+    }
+
+    private function createIndexWithoutAlias($alias)
```
Comment (mhujer) [inline tests/Elasticsearch/Tests/ClientIntegrationTests.php]: there can be a `string` typehint

File: tests/Elasticsearch/Tests/ClientIntegrationTests.php [hunk 7]
```diff
@@ -18,9 +19,32 @@
  */
 class ClientIntegrationTests extends \PHPUnit\Framework\TestCase
 {
+
+    /**
+     * @var $client Elasticsearch\Client
+     */
+    protected static $client;
```
Comment (mhujer) [inline tests/Elasticsearch/Tests/ClientIntegrationTests.php]: What about making it non-static a recreating it for each test in `setUp()`? (it probably does not matter here, but I think it is a good practice to keep tests independent)

Comment (Nexucis) [reply tests/Elasticsearch/Tests/ClientIntegrationTests.php]: well it's just to not initialize the client in every test below. I'm not sure to understand why it makes the test dependent. If you don't want to use the global client, you can ever initialize a new one.

File: tests/Elasticsearch/Tests/ClientIntegrationTests.php (hunk 3 above)
Comment (Nexucis) [reply tests/Elasticsearch/Tests/ClientIntegrationTests.php]: thanks for the tips, i didn't know that :)

File: tests/Elasticsearch/Tests/ClientIntegrationTests.php (hunk 5 above)
Comment (Nexucis) [reply tests/Elasticsearch/Tests/ClientIntegrationTests.php]: i'm sorry, i don't understand what do you mean by that :s

Comment (mhujer) [reply tests/Elasticsearch/Tests/ClientIntegrationTests.php]: You can change the method signature to `private function createIndexWithAlias(string $alias)` (added `string`), so the parameter type is checked.

File: tests/Elasticsearch/Tests/ClientIntegrationTests.php (hunk 7 above)
Comment (Nexucis) [reply tests/Elasticsearch/Tests/ClientIntegrationTests.php]: oh sorry, I read your comment in a wrong way. Well yes we can initialiaze it in a `setUp()` method

File: tests/Elasticsearch/Tests/ClientIntegrationTests.php (hunk 5 above)
Comment (Nexucis) [reply tests/Elasticsearch/Tests/ClientIntegrationTests.php]: thanks for the explanation :)

**Comment by Nexucis** [comment]
These new commit should fixe the review made by @mhujer (thanks for the review by the way)

File: tests/Elasticsearch/Tests/ClientIntegrationTests.php [hunk 8]
```diff
@@ -18,9 +19,29 @@
...
+        $param = [
+            'index' => '_all'
+        ];
+        $this->client->indices()->delete($param);
+    }
+
     public function testCustomQueryParams()
     {
-        $client = Elasticsearch\ClientBuilder::create()->setHosts([$_SERVER['ES_TEST_HOST']])->build();
+        $this->client = Elasticsearch\ClientBuilder::create()->setHosts([$_SERVER['ES_TEST_HOST']])->build();
```
Comment (mhujer) [inline tests/Elasticsearch/Tests/ClientIntegrationTests.php]: I think this line can be removed completely, because `$this->client` is already initialized in `setUp()`

Comment (Nexucis) [reply tests/Elasticsearch/Tests/ClientIntegrationTests.php]: luckily you're here, I completely forgot to remove it. Thanks again

**Comment by Nexucis** [comment]
ping @polyfractal :)

**Comment by Nexucis** [comment]
pong @polyfractal 😏

**Comment by Nexucis** [comment]
For an undkown reason, the test don't pass with `docker.elastic.co/elasticsearch/elasticsearch:6.0.0`. Which seems different here, because all test pass. 

So I don't  if there is a problem with the docker image or with the current ci here.

Maybe it's another reason to consider the issue #693 (i'm kidding)

**Comment by Nexucis** [comment]
ping @polyfractal and Happy new Years :)

**Comment by ezimuel** [comment]
Closing because it's related with #482.
